### ISSUE TITLE
Not able to deploy contract on network 

### ISSUE THREAD

**Issue by romisalve** [trufflesuite/truffle#1997]
## Issue

I have developed a Quorum maker network and configured the truffle-config.js to connect to it. If I open truffle console, it makes no errors, so I understand is accessing the network. Also, I am able to acces web UI of quorum maker so in fact the network is created. 

However, when I tried to deploy a contract through truffle, I got: 
![image](https://user-images.githubusercontent.com/21060405/57374519-746c2b80-7171-11e9-843f-e4825474acc1.png)

Also, if I deploy the contract directly from quorum maker web UI, the contracts appears to be deployed but if form truffle console I run "networks", I get that no contract is deployed on the network, so something is generating that inconsistency and I don't understand the reason. 

## Steps to Reproduce
Create Quorum maker network on centOS virtual Machine with IP 192.168.26.9 and RPC port 22000
Join node on same IP but different ports
configure truffle-config.js to work with that network
Compile contracts
Try to deploy

## Expected Behavior

Get my contract deployed and accessible over truffle console and stop getting inconsistencies with quorum maker

## Actual Results
 This error got displayed: 
![image](https://user-images.githubusercontent.com/21060405/57375040-bf3a7300-7172-11e9-8b92-fb706023e2e6.png)

And truffle console shows (the first output is about ganache, the second one is the one of Quorum Maker)
![image](https://user-images.githubusercontent.com/21060405/57375075-de390500-7172-11e9-962d-79712d24f0e8.png)

## Environment

* Operating System: CentOS 7
* Ethereum client: Geth - Quorum Maker network 
* Truffle version (`truffle version`):  v5.0.14
* node version (`node --version`): v10.15.1
* npm version (`npm --version`): 6.4.1

**Comment by eggplantzzz** [comment]
Can you show us your truffle config?

**Comment by romisalve** [comment]
I commented the lines of "from" and "websockets" because when I include "websockets=true" I only got "connection not open send()"
I am able to open truffle console on development, which makes me think it is identifying some network, but I ran "web3.eth.isSyncing()" and got false. I also got false on "web3.eth.isMining()" . Also, on "web3.eth.defaultAccoun" I got null eventhough I set the "from" field on truffle-config.js
![image](https://user-images.githubusercontent.com/21060405/57465255-98ed0400-7254-11e9-9f45-f876e1470340.png)

**Comment by romisalve** [comment]
Also when I ran the command to get the provider, I got 
![image](https://user-images.githubusercontent.com/21060405/57465397-e9646180-7254-11e9-8373-8939899291b3.png)

**Comment by eggplantzzz** [comment]
Hmm yeah it seems like there is something wrong with the way you are connecting. I'll talk to some of my team that are more familiar with quorum.

**Comment by eggplantzzz** [comment]
But that error you are getting, "authentication needed: password or unlock" is not from Truffle. Or actually you probably need to unlock some accounts on starting Geth so that they can be used.

**Comment by romisalve** [comment]
I tried to get around that error by setting "from" on deployer with one of the 2 addresses. I put the one I was able to unlock which happens to be the one with balance. 
The other account, which is coinbase, was impossible to unlock. It asks for a passphrase, which is supposed  to be an empty string but introducing an empty string fails.

I managed to unlock one account and I reach this point: 
![image](https://user-images.githubusercontent.com/21060405/57544271-47fd0e80-732d-11e9-8357-5114a962f1b3.png)

It seems that migrations contract got deployed, and actually on quorum maker web UI it appears that one contract was deployed. But I do not understand why it keeps asking for authentication.

**Comment by romisalve** [comment]
Also I do not understand why can the quorum maker web UI showing that there is a contract that got deployed, but is not shown on truffle console. Also, even though the contract is deployed, the ABI is not upload. On Quorum wiki it says that a symlink needs to be created, but from where to where?

**Comment by eggplantzzz** [comment]
You might try unlocking the accounts on the Geth side of things. In the past I have always created the accounts and unlocked them as part of the startup command. Perhaps it is different with Quorum...I'll ask my colleague who has more Quorum experience.

Well, like the message says, the network state is unknown so maybe one of the transactions was completely successful and then failed for some other reason. I'm not sure :)

**Comment by romisalve** [comment]
I believe that the first reason of failure is the first error of authentication. 

Have you been able to ask your colleague? Or do you have any idea on the question about the symlink or reason for the inconsistency between truffle and Quorum maker?

**Comment by CruzMolina** [comment]
Hey @romisalve ! I suspect there's some kind of miscommunication happening between Quorum Maker and Truffle. Will need some time to explore Q Maker under the hood, but I have a hunch that unlike the Quorum 7nodes example (https://github.com/jpmorganchase/quorum-examples/tree/master/examples/7nodes), Q Maker doesn't handle minting of new blocks. It _looks_ like Q Maker confirms tx receipt, but that the pending block is never minted.

If you have more insight into how Q Maker handles block validation, would love to hear about it 🙏

**Comment by CruzMolina** [comment]
@romisalve , following up on this. There are a couple of things going on.

1) websockets definitely aren't working w/ Quorum, but I'm not sure if it's something that needs to be enabled on the Quorum Maker side, or if it's the version of `web3.js` we're using that's currently unable to converse with Q Maker using websockets. 
1) Since Quorum is a permissioned chain, we can expect standard `web3.js` methods like `web3.eth.isSyncing` & `web3.eth.isMining` to return false (no mining, only validating). Also, the web3 lib we're currently using (`web3.js`) [especially versions >1.0] is very EVM-specific.
1) There appears to be a default timeout occurring on the Q Maker side of things. If I `migrate` and `migrate --reset` a bunch within roughly 5 mins of joining/creating a Q Maker network, everything works as expected. However, there _could_ be something we're not handling on our side of things.

**Comment by romisalve** [comment]
Actually I have not changed anything on my configuration and now is deploying. The only thing I have to do is unlock the node account. 
Well, I also had to split my contract in smaller ones because it said that the contract could not be store, but once I split it, I was able to deploy to the network and from truffle console, it shows that a contract is deployed. 
Anyway, on quorum maker web UI there are no ABI for the contracts. 
Do I need to have the ABI accesible through quorum maker web UI or it is not necessary and I can work directly with truffle if I want to develop  a dApp?

**Comment by CruzMolina** [comment]
@romisalve How were you able to unlock the geth node perpetually? It looks like Q Maker uses an older version of `geth` that doesn't have the `--allow-insecure-unlock` flag so the node appears to lock up after the default 300s.

What was your error for the large contract? IIRC, Quorum has a super large block gas limit so I'm surprised you ran into that type of error. Can you share a link to your contract?

Regarding the ABI, I'm not sure yet on the scope of Q Maker, but if you're able to provide reproducible steps (i.e. link to a repo) I can take a look.

**Comment by romisalve** [comment]
No, I was not able to unlock it perpetually, but web3.personal.unlockAccount("address", "pass", 0); unlocks it until geth client is shut down. 

I can share my contract of course, can I paste it here or how do you want me to share it? I don´t understand de reason of that error. 

Now I tried to split it into smaller contracts and I am getting this error when trying to call a function but I don´t understand what is happening on my contract 
![image](https://user-images.githubusercontent.com/21060405/57934808-d4ac4d00-7896-11e9-9168-7bd06ee73e6f.png)

**Comment by CruzMolina** [comment]
Can you share a repo link to your truffle project? Would make it easier for reproduction and debugging.

Where & when are you making the `web3.personal.unlockAccount` call?

**Comment by romisalve** [comment]
https://github.com/romisalve/Quorum-POC.git There is my contract. There are a lot of commented lines due to the following reasons :

When I tried to compile it I got "stack too deep error" and started commentig lines of returns (of course it was not the most intelligent decision but I didn´t how to solve it)
I have a lot of internal functions which actually I believe I need them as public because as internal they can only be called inside the contract or I am wrong? But if I put them as public, I end on the same error of "can´t store.." 

Now I was trying through truffle console (from where I made the web3.eth.personal.unlockAccount and I was being able to deploy that contract). I was trying this: 

CrudAEOCreate.deployed().then(function(instance) { crud = instance;})

 crud.createMasterDataAEO(1234, "primerFuncRefNum1234", "primerDocName", "receipType", "receiptName", "senderType", "senderAEO", "12456" );

crud.associateMasterDataParty("primerFuncRefNum1234","PrimerpartyNameAEO","PrimerShortNameAEO","PrimerBusinessTypeAEO",2601,"Peru","PrimerRoleCodeAEO");

 crud.readAEO( "primerFuncRefNum1234","senderAEO","PrimerpartyNameAEO", 2601, "PrimerRoleCodeAEO");

And the two first don't give any error, but the third one gives the error of the image on the previous post.  The first two gave this outputs: 

![image](https://user-images.githubusercontent.com/21060405/57937231-e17f6f80-789b-11e9-8a26-a556c379ed82.png)

![image](https://user-images.githubusercontent.com/21060405/57937257-f0662200-789b-11e9-83ef-f6abd8b98d16.png)

On my contract, on the readAEO function, actually I need to have this on return: 
returns(string endDateAEO, string memory recipientAEO,
                                    string memory partyShortNameAEO, string memory businessTypeAEO, string memory identificationIssuingCountryAEO)

But with that statement, when I tried to call the function I got " returned values aren't valid, did it run out of gas?" so I try changing the return to an uint but anyway I am not being able to execute the function.

**Comment by romisalve** [comment]
And also, do you know from where can I read what does each field of those outputs represent?

**Comment by CruzMolina** [comment]
Thanks for the link!

>When I tried to compile it I got "stack too deep error"

This _might_ be resolved in the latest truffle release (5.0.18).

Regarding your `internal` functions, yes if you want them to return publicly something while on-chain then they need to be `public` (I'll take a look shortly to scope out your contracts functionality).

>But with that statement, when I tried to call the function I got " returned values aren't valid, did it run out of gas?"

I suspect this is because some of your transactions are private. I'll check it out.

**Comment by CruzMolina** [comment]
Uh oh, is your repo private? It's showing up as not found.

**Comment by romisalve** [comment]
Uh, it was private. Now I made it public , can you access?

**Comment by CruzMolina** [comment]
Got it. Thanks! Will take a look in a bit 👍

**Comment by romisalve** [comment]
Thank you!!

**Comment by romisalve** [comment]
were you able to look in?

**Comment by CruzMolina** [comment]
Hey @romisalve , the stack too deep error I ran into was `solc-js` specific:

```
truffle compile                 

Compiling your contracts...
===========================
> Compiling ./contracts/CrudAEOCreate.sol

/Users/cruzmolina/Code/truffle-projects/example/Quorum-POC/contracts/CrudAEOCreate.sol:207:22: CompilerError: Stack too deep, try removing local variables.
    if(addressExists[partyIdAEO]){
                     ^--------^

Compilation failed. See above.
Truffle (unbundled) (core: 5.0.18)
Node v11.4.0
```

I was able to comment out what appears to be redundant code and some unused variables to get it to work.

**Comment by CruzMolina** [comment]
Just opened a PR on your repo, let me know if compilation works on your end 🤞

**Comment by CruzMolina** [comment]
Also, regarding your `crud.readAEO` call in `truffle console`, if you notice in my PR, I commented out 3 unused parameters in the `readAEO` function definition. You were probably receiving the ` returned values aren't valid` error, because currently `readAEO` doesn't do anything with those 3 parameters.

If you test the code in my PR in the console and instead only pass something like `crud.readAEO( "primerFuncRefNum1234", "PrimerpartyNameAEO");`, you should receive appropriate output w/o an error.

**Comment by romisalve** [comment]
@CruzMolina I am really thankful for your help and I have some questions if it is possible. 
I tried to deploy the contract with the changes of your PR but I am facing the error that contract code could not be stored, how can I solve this?
![image](https://user-images.githubusercontent.com/21060405/58103349-34696780-7bb9-11e9-9615-3184a5e345fc.png)

In the read function may I add the return statement with some string output parameters and some uints? 

Also, at the beginning of the contract I create some arrays and mappings. Originally, the contract was a CRUD contract but due to this same error, I decided to split it up into smaller contracts. What happen with those arrays and mapping created on the first contract if I need them for my operations on the other contracts? If I deploy all the contracts on my network, are they reachable from other contracts where they are not defined? Or which would be te approach?

And a final question, how can I expose my smart contract operations in order to consume them but no directly from truffle? I was planning to use Camunda BPM modeler to model some forms and consume smart contract methods. It is possible?

**Comment by CruzMolina** [comment]
@romisalve What does your truffle config look like?

What worked on my end was:

```
module.exports = {
  networks: {
    nodeone: {
      host: "127.0.0.1",
      port: 22000,
      network_id: "*",
      gasPrice: 0,
      gas: 450000000,
      type: "quorum"
    }
  }
};
```

Also, if you could put up your entire truffle project in that repo, it would help me reproduce more easily.


i.e. truffle project layout
```
├── contracts
│   ├── CrudAEOCreate.sol
│   └── Migrations.sol
├── migrations
│   └── 1_initial_migration.js
├── test
└── truffle-config.js
```

**Comment by CruzMolina** [comment]
>In the read function may I add the return statement with some string output parameters and some uints?

That should be fine. IIRC, the problem with the read calls you were making were directly related to your read function definition in the smart contract. The definition had multiple parameters that were completely unused or inaccessible. That's why web3 was throwing that "out of gas?" returned value err. I _think_ as long as you make sure whatever parameters are being passed are used in some manner within the function, you should be fine.

>What happen with those arrays and mapping created on the first contract if I need them for my operations on the other contracts? If I deploy all the contracts on my network, are they reachable from other contracts where they are not defined? Or which would be te approach?

Any mappings that are public can be read from another contract.  I believe if you want to add to those mappings, they either need to be added via a method in that contract, or through a contract you deploy that inherits that contract.

>And a final question, how can I expose my smart contract operations in order to consume them but no directly from truffle? I was planning to use Camunda BPM modeler to model some forms and consume smart contract methods. It is possible?

Oh, that's a really cool idea. You can probably use the Camunda BPM modeler with the contract json artifacts found after compilation in the `build` folder.

You might want to take a look at the abi spec (https://github.com/trufflesuite/truffle/tree/develop/packages/truffle-contract-schema#abi) since you're thinking specifically about contract methods. On that page you can also find the general contract json object schema although it needs to be updated since we now support NatSpec in Solidity contracts (https://solidity.readthedocs.io/en/develop/natspec-format.html). Using NatSpec you can have `userdoc` and `devdoc` auto-populated into your contract build artifact. That might be a better higher level choice than using the abi. This documentation is a more up-to-date higher level look at how contract object json's get defined and written: https://github.com/trufflesuite/truffle/tree/develop/packages/truffle-artifactor#artifactorsavecontractdata

**Comment by romisalve** [comment]
> That should be fine. IIRC, the problem with the read calls you were making were directly related to your read function definition in the smart contract. The definition had multiple parameters that were completely unused or inaccessible. It looks like that's why web3 was throwing that "out of gas?" returned value err. I _think_ as long as you make sure whatever parameters are being passed are used in some manner within the function, you should be fine.

Thank you!! I'll have that in mind for every function.

> Any mappings that are public can be read from another contract. I believe if you want to add to those mappings, they either need to be added via a method in that contract, or through a contract you deploy that inherits that contract.

Ok, great. And to inherit a contract I need to define it as "_ contract nameOfTheContract is NameOfTheContractIWantToInheritFrom _ " and that is it? Then, to deploy them, in migrations directory I need to have a file for each contract or I can deploy them all thorugh the same migration file?


> Oh, that's a really cool idea. You can probably use the Camunda BPM modeler with the contract json artifacts found after compilation in the `build` folder.
> 
> You might want to take a look at the abi spec (https://github.com/trufflesuite/truffle/tree/develop/packages/truffle-contract-schema#abi) since you're thinking specifically about contract methods. On that page you can also find the general contract json object schema although it needs to be updated since we now support NatSpec in Solidity contracts (https://solidity.readthedocs.io/en/develop/natspec-format.html). Using NatSpec you can have `userdoc` and `devdoc` auto-populated into your contract build artifact. That might be a better higher level choice than using the abi. This documentation is a more up-to-date higher level look at how contract object json's get defined and written: https://github.com/trufflesuite/truffle/tree/develop/packages/truffle-artifactor#artifactorsavecontractdata

Ok, so there is no way to have kind of an URL in order to hit it from POSTMAN for example and consume it services, isn´t it?  So I need to have like kind of a Java class for example from which I use the artifacts generated or how should it be done??

And in relation with this, If I am planning to have a network with four nodes. Do I need to have in each node that is going to interact with the network truffle installed?

**Comment by romisalve** [comment]
@CruzMolina I put up my entire truffle project. The account on the migrations file is the account of the node to which I am accessing through truffle.

**Comment by CruzMolina** [comment]
Hmm . I still seem able to deploy successfully locally using your truffle project.

**Comment by CruzMolina** [comment]
Perhaps you need to up your gasLimit (`gas` key in your truffle config)? How are you configuring and running your Q Maker network?

**Comment by CruzMolina** [comment]
>to inherit a contract I need to define it as "_ contract nameOfTheContract is NameOfTheContractIWantToInheritFrom _ " and that is it? Then, to deploy them, in migrations directory I need to have a file for each contract or I can deploy them all through the same migration file?

Yes, `contract NewContract is InheritedContract`. You don't necessarily need each contract in a different migration file. For example, you can have your second migrations file be `2_deploy_contracts.js` and within that file module export deploy, transact with, and/or link your contracts.

> Ok, so there is no way to have kind of an URL in order to hit it from POSTMAN for example and consume it services, isn´t it? So I need to have like kind of a Java class for example from which I use the artifacts generated or how should it be done??

Not currently, although I believe there are a number of GraphQL-like efforts to make this happen. With Camunda BPM's modeler it looks like you can define a json schema and then feed it the artifact files. I'm not sure though, because I've never played around with it. You wouldn't need truffle on each of those nodes. Just the build artifacts either hosted somewhere accessible or on each node.

**Comment by romisalve** [comment]
@CruzMolina 
> Perhaps you need to up your gasLimit (`gas` key in your truffle config)? How are you configuring and running your Q Maker network?
The truffle-config.js file I put on my project has been modified with your suggestion (before that I was having less gas and couldn't store the contract)

> Yes, `contract NewContract is InheritedContract`. You don't necessarily need each contract in a different migration file. For example, you can have your second migrations file be `2_deploy_contracts.js` and within that file module export deploy, transact with, and/or link your contracts.
Ok, thank you! I'll try that later today. 

> Not currently, although I believe there are a number of GraphQL-like efforts to make this happen. With Camunda BPM's modeler it looks like you can define a json schema and then feed it the artifact files. I'm not sure though, because I've never played around with it. You wouldn't need truffle on each of those nodes. Just the build artifacts either hosted somewhere accessible or on each node.

Ok, thank you!. And do you have any example (no matter the programming language) of how is this achieved?  I am having trouble on understanding which is the work around. I would need to create kind of a client to interact with truffle artifacts?

Thank you a lot!!

**Comment by romisalve** [comment]
@CruzMolina 

I have try this two ways to deploy on the migration file: 
` 
```
var CrudAEOCreate = artifacts.require("./CrudAEOCreate.sol");
var CrudAEOUpdateAndDelete = artifacts.require("./CrudAEOUpdateAndDelete.sol");

module.exports = function(deployer){

        deployer.deploy(CrudAEOCreate,{from: "0x1caFCe8fF232515Dcb07eE7901D49cFF8397eF7c"}).then(
  () =>  deployer.deploy(CrudAEOUpdateAndDelete,{from: "0x1caFCe8fF232515Dcb07eE7901D49cFF8397eF7c"});


};
```
   And 

```
var CrudAEOCreate = artifacts.require("./CrudAEOCreate.sol");
var CrudAEOUpdateAndDelete = artifacts.require("./CrudAEOUpdateAndDelete.sol");

module.exports = function(deployer){

        deployer.deploy(CrudAEOCreate,{from: "0x1caFCe8fF232515Dcb07eE7901D49cFF8397eF7c"});
deployer.deploy(CrudAEOUpdateAndDelete,{from: "0x1caFCe8fF232515Dcb07eE7901D49cFF8397eF7c"});


};

```
But both ways ends up in this error: 
![image](https://user-images.githubusercontent.com/21060405/58185219-ffc4e100-7c88-11e9-8448-755a006948f9.png)

**Comment by CruzMolina** [comment]
>And do you have any example (no matter the programming language) of how is this achieved? I am having trouble on understanding which is the work around. I would need to create kind of a client to interact with truffle artifacts?

It depends on your development/production environment. If you're just running your nodes locally, I imagine you could spin up a local server to host the artifact files or host and feed the artifact to a modeler. Again, haven't really played around with modeling but I imagine that's the high-level extent of what you want to do.

Regarding your inherited contract migrations, you should only have to deploy contract B (since it inherits contract A). You might want to check the ordering of your contract definitions in your source file. Make sure contract A is defined first. Feel free to push your project to the prior shared repo if you'd like me to take a look.

>in which cases I should define my functions as internal?

Any function you only want your deployed contract to be able to access within another function in that same contract (or derived contracts) should be defined `internal`. Feel free to check out the Solidity docs for more info (https://solidity.readthedocs.io/en/v0.5.8/types.html#function-types & https://solidity.readthedocs.io/en/v0.5.8/solidity-by-example.html#modular-contracts)

**Comment by romisalve** [comment]
> It depends on your development/production environment. If you're just running your nodes locally, I imagine you could spin up a local server to host the artifact files or host and feed the artifact to a modeler. Again, haven't really played around with modeling but I imagine that's the high-level extent of what you want to do.

I have 4 nodes, each of them on a separate machine. They are going to be the 4 nodes of the quorum network. So I should have a local server to host the files?  What I am not being able to understand is in which way I should use those files of build directory. I don't understan how I would end up calling one of the functions. 


> Regarding your inherited contract migrations, you should only have to deploy contract B (since it inherits contract A). You might want to check the ordering of your contract definitions in your source file. Make sure contract A is defined first. Feel free to push your project to the prior shared repo if you'd like me to take a look.

I also tried to deploy only contract B but is still failing. I am going to push it to my repo, thank you a lot!! The contracts can be in different files or they need to be in the same file? 



> Any function you only want your deployed contract to be able to access within another function in that same contract (or derived contracts) should be defined `internal`. Feel free to check out the Solidity docs for more info (https://solidity.readthedocs.io/en/v0.5.8/types.html#function-types & https://solidity.readthedocs.io/en/v0.5.8/solidity-by-example.html#modular-contracts).

Ok, great! I was doing this on some of my new functions.

**Comment by romisalve** [comment]
I was missing the import sentence at the beginning of my file.I am going to see if it works now and then I'll push it to my repo.

**Comment by CruzMolina** [comment]
Contracts can be in different files, but you'll need to remember to import. If they are in the same file, I believe the inherited contract needs to be defined before the derived contract:

```
Contract A {
 ...
}

Contract B is A {
 ...
}
```

**Comment by CruzMolina** [comment]
>What I am not being able to understand is in which way I should use those files of build directory

Ah, I think I misunderstand how the modeler works. It looks like you'll need to host a server with the build artifacts and either leverage `web3` or `truffle-contract` to consume the artifacts on that server before possibly using the modeler to make function calls. It looks like Camunda has a js pkg to handle some of this. Again, not too familiar with this, so CCing @gnidan in case he has more insight.

**Comment by romisalve** [comment]
@CruzMolina Thank you!
Yes, I found that. After importing it I was able to compile them. I pushed them to the repo :) I am going to give a look to the read function and I am going to set the return parameters to what I need them to be and see if it works. 

Then I am going to have a look at what you are saying about leveraging web3 and truffle contract.  But when you say leverage web3 or truffle contract on a server, that is separate of the truffle on my node, isn´t it?  And then how would be the interaction between modeler-server-truffle? 
Sorry maybe I am getting confused with something very simple but I don't see how to begin with that part.

**Comment by CruzMolina** [comment]
>But when you say leverage web3 or truffle contract on a server, that is separate of the truffle on my node, isn´t it?

Yes, those are separate nodejs packages. I imagine you'd have a local server running that has access to the build artifacts, the web3 and/or truffle-contract pkg's installed and accessible in the backend, along with the modeler nodejs pkg. I _think_ you should be able to string those parts together and run the modeler against your exposed contract methods (exposed via truffle-contract or web3.Contract).

**Comment by romisalve** [comment]
@CruzMolina 
Ok, I believe I got what you said. Once I reach that point I will investigate more on it, thank you very much! But some questions on this.
Why do I need to have it on a local server on a different machine than the one running the node with truffle? Is for something related to security of some kind? And once I have that up, how it distinguish if I am accessing from which node? Or it is like one truffleContract for each node to access the build artifacts? That is what I don't have completely clear until know. 
I have four nodes, I should have one server with truffle contract to access the artifacts, but access them from which node? Or maybe that is defined in some place? If I want to create something from nodeOne and read it from ndoeFour, everything is accessing artifacts through the same truffleContract or web3.Contract server? And why it should be another machine apart from the ones of the four nodes and only one? Maybe I am not being clear on my questions and I should rephrase them, sorry. 

I was trying to deploy my contracts on my network and now that I add the new contract I am facing  the "contract couldn't be stored, please check your gas limit" again and nothing has changed on my truffle-config.js.  
When you inherit a contract all the inherited contract code gets copied to the derived one?  Because if this is what is causing the "contract code couldn't be stored,.." then which would be the workaround to this problem?  I don't need to have access to all CrudAEOCreate.sol functions on CrudAEOUpdateAndDelete.sol contract, maybe there is a way to import only the necessary stuff. 
What I definitely need are the arrays and mappings created on the first contract.  Could you deploy it? Maybe you haven't had time, sorry.

**Comment by CruzMolina** [comment]
>Why do I need to have it on a local server on a different machine than the one running the node with truffle?

It really depends on how you plan to set the network up in production. I imagine you'd want to have the artifacts consumed and available on at least one machine where a node is running. I'm not sure about the scope of what a modeler can accomplish, but if you're looking to use it to more easily abstract and actually call contract functions, then I imagine as long as you have one machine hosting the artifacts with the modeler consuming them and with that modeler accessible, then you would only need one server making calls to one node that in turn communicates and updates all the other nodes.

I'll take a look again at your repo in a bit. At a glance, it looks like you've updated it, but in your second migrations file I believe you're deploying the same contract twice.

Regarding this:
> I don't need to have access to all CrudAEOCreate.sol functions on CrudAEOUpdateAndDelete.sol contract, maybe there is a way to import only the necessary stuff.

You could possibly do this by modularizing your base contract into separate smaller contracts. That way when you define the complete contract to be deployed, it only inherits parts that are actually used.

**Comment by romisalve** [comment]
And so there is no way to define a sentence that only imports certain stuff of a contract? I need to modularize it on smaller ones?

**Comment by CruzMolina** [comment]
>And so there is no way to define a sentence that only imports certain stuff of a contract? I need to modularize it on smaller ones?

https://solidity.readthedocs.io/en/v0.5.8/layout-of-source-files.html?highlight=import#importing-other-source-files

Looks currently you will need to modularize things.

>So I could have a contract creating the arrays and mappings and inherit it on each of the other contracts and all of them would be accessing to the same stored values

I believe that should work!

> I want to show that if a create an entity through one node, then if I go to another one and execute the function of read, then I would obtain that entity created by the other node.

If you are trying to physically represent this in some manner, it probably makes sense to have two separate machines hosting the artifacts and able to represent the contracts with access to the network state. The trickiest part will probably be stringing together the modeler with web3 and/or truffle-contract to execute a function on one machine and then read the updated contract state on the other.

**Comment by romisalve** [comment]
@CruzMolina  thank you! I'll modularize it and once I try to deploy it, I'll update my repo. 

> If you are trying to physically represent this in some manner, it probably makes sense to have two separate machines hosting the artifacts and able to represent the contracts with access to the network state. The trickiest part will probably be stringing together the modeler with web3 and/or truffle-contract to execute a function on one machine and then read the updated contract state on the other.

- This two machines you mention in which I would be hosting the artifacts, can they be the same machines that are running the nodes? 
- The artifacts of the contracts can be generated by truffle on one machine and then I can copy that directory to the other machines that run the other nodes?
- It is mandatory to have the json files that contains the contracts artifacts on each node of the network from which I intend to interact with the network? 
- So, I should have as many machines with the modeler (or not necessary a modeler, maybe just an entity which is supposed to consume the smart contract functions) and truffle-contract/web3 as nodes from which I am planning to access the network? 
- Why are you saying it would be so tricky to do that? Once a function is executed isn't it kept on the blockchain which is the same all over the network? Maybe I'm aiming to do something that is not so achievable, but I am having trouble to understand the reason. 

Thank you a lot!!

**Comment by CruzMolina** [comment]
>This two machines you mention in which I would be hosting the artifacts, can they be the same machines that are running the nodes?

I don't see why not!

>The artifacts of the contracts can be generated by truffle on one machine and then I can copy that directory to the other machines that run the other nodes?

Yes, or at least make sure the machine w/o the artifacts locally has access to the hosted artifacts.

>It is mandatory to have the json files that contains the contracts artifacts on each node of the network from which I intend to interact with the network?

No, only whichever machine(s) you plan to run the modeler on.

>So, I should have as many machines with the modeler...

I imagine a simple demo would be two separate machines running the modeler to visualize interaction with contracts on the blockchain. If you'd like to expand it, then you'd need to make sure each additional machine has access to the artifacts, the modeler js pkg & web3/truffle-contract.

>Why are you saying it would be so tricky to do that? Once a function is executed isn't it kept on the blockchain which is the same all over the network?

IMO, the tricky part is stringing it all together. I think it's do-able and might end up being pretty straightforward. The nodes store the bytecode, (IIRC) they have no concept of the source code or smart contract methods. This is why the artifacts, and web3 and/or truffle-contract are important because they help make sense of the code that's been deployed and what's happening on-chain. The major paint point here is needing to have this "full stack" of blockchain-specific components (nodes, contract artifacts, contract instances consuming the artifacts via web3/truffle-contract, etc) on every separate machine you'd like to demo and model.

**Comment by romisalve** [comment]
@CruzMolina  I have another question more related to programming in solidity itself. 

Solidity limits local values on functions to a maximum of 16 (including parameters and return values), doesn't it? Due to the fact that the AEO I intend to create on my contract has too many attributes, I had to spilt the create function into smaller functions that creates  all the different "pieces" of the AEO and associate each of them to the corresponding MasterData entity.  

The workflow I am thinking on is that once the user has completed all the necessary fields to create an AEO and press the submit button, each of the functions will get called with its corresponding parameters. 

The problem I am having is the following. Every executed function represents a transaction on the network, doesn't it?  Is there a way to ensure that a function only gets called once the previous one succeed? So the transactions are sequentially? 

Anyway, this wouldn't get my problem solved. The problem I am having is: what happens if for example, the first three transactions get executed but the fourth one fails? Then the AEO would be on half way of being created but those transactions can't be reverted, what can be done? Maybe is something that can be added to the way I am programming my functions or I don´t know which would be a walkaround this.  Or simply those transactions will be stored with no purpose on the blockchain? 

Is there a way to keep track that certain transaction, result of certain function with certain parameters, have been already made? Because if certain user tries to create the same AEO that wasn't created entirely as I exemplify on the previous paragraph , all the values created on the first attempt with the functions that were executed successfully have to be created all over again? 

Thank you a lot! I will read your previous answer, I haven't notice you answered me.

**Comment by eggplantzzz** [comment]
@romisalve Regarding making transactions in chronological order, the javascript methods that you use with your contract objects return Promises. So you can use `then` syntax or `async/await` syntax to ensure that the second method gets called only after the previous one is finished.

**Comment by CruzMolina** [comment]
>The problem I am having is: what happens if for example, the first three transactions get executed but the fourth one fails? Then the AEO would be on half way of being created but those transactions can't be reverted, what can be done? Maybe is something that can be added to the way I am programming my functions or I don´t know which would be a walkaround this. Or simply those transactions will be stored with no purpose on the blockchain?

It's been a while but there's a way to do this within your contracts. You might need to dig around in the Solidity docs some more or perhaps look at a couple of their examples. IIRC, what you do is wrap your `public` method calls around `internal` functions and possibly a `require` statement.

So, say you have a method:

```function executeSequentialTx() public returns (bool) { }```

you have each separate method as an `internal` function like...

```
function firstTx() internal returns (bool) {
  ...
  return True;
}
```

and then you wrap `executeSequentialTx` like so:

```
function executeSequentialTx() public returns (bool) {
  require(firstTx() && secondTx() && thirdTx());
}
```

something along those lines. I think there's a number of ways to do it. Again, this is much more Solidity-related than Truffle-proper and it's been a while.

There's also a difference between using a `require` statement and `assert`. I think a require, if it fails, will revert and refund the gas, while an `assert` will fail and lose the gas?

Anywho, hope that helps steer you in the right direction.

>Is there a way to keep track that certain transaction, result of certain function with certain parameters, have been already made? Because if certain user tries to create the same AEO that wasn't created entirely as I exemplify on the previous paragraph , all the values created on the first attempt with the functions that were executed successfully have to be created all over again?

See the above. You can also create a mapping to map users and their AEO creations. i.e. a `modifier` check to ensure a user can only create an AEO if they haven't already done so (and before the end of your AEO creation you ensure that the user's address is added to the mapping).

**Comment by romisalve** [comment]
@eggplantzzz thank you! I will take that into account when calling my functions. 

@CruzMolina thank you! Yes, require refund the gas while assert does not. Then, about using truffle- contract: I was trying to use it and try to consume a function (createMasterDataAEO). For that reason, I created an app.js file in which I put: 

```
const Web3= require('web3');
const truffleContract= require('truffle-contract');

const web3= new Web3.providers.HttpProvider("http://192.168.26.9:22000");
const ContractJSON= require ('./build/contracts/CrudAEOCreateAndRead.json');

var MyContract = truffleContract(ContractJSON);

MyContract.setProvider(web3.currentProvider);

MyContract.deployed().then(function(instance) {

        return instance.createMasterDataAEO(1234, "primerFuncRefNum260196", "primerDocName", "receipType", "receiptName", "senderType", "senderAEO", "12456" );}).then(function(result) {
        console.log(result);
}).catch(err=> {console.log(err);});

```

This would be a proper approach? I am creating a file inside my truffle directory (I mean, that file is in the same directory that contracts directory, build directory, and so on).  This would be kind of the way to consume the smart contract function through my modeler? 

I have not been able to try it yet because I have not been able to get truffle-contract installed. I have tried npm install truffle-contract (and same with sudo) but I didn't succeed.  

The const web3 is what is going to change for each node where I am going to install truffle-contract, ins't? And the ContractJSON actually needs to be define for every contract I want to interact, and for each node? 

By copying those .json files into each node, I assure that they are working through the same contract instance, don't I? Because I must be acting on the same structures, otherwise it would be pointless 

Thank you a lot! Sorry the large amount of questions, I am working on this alone and that's why I find this support so important and I'm completely grateful.

**Comment by romisalve** [comment]
@CruzMolina If I am using truffle console to try my smart contract functions, how can I show on console the boolean that, for example, my public function "AssociateMasterDataParty" should be retrieving?

I tried to use .call() but it says that is not a function. 
I tried to assign the execution to certain variable, as result= instance.MyFunction() and then print reuslt, but it gives me the transaction receipt, and if I run result.ToString() it retrieves "Objetc Promise". 

Is this achievable or my function should be defined as constant in order to be able to show it return value? 

Thank you!

**Comment by CruzMolina** [comment]
>This would be a proper approach? I am creating a file inside my truffle directory (I mean, that file is in the same directory that contracts directory, build directory, and so on). This would be kind of the way to consume the smart contract function through my modeler?

I'm not exactly clear on the scope of the modeler js pkg, but I _think_ whatever button you'd like to visually click to interact with your contract and the blockchain should interact with `truffle-contract` and/or `web3` to actually make the call. The created contract instance should either receive the response and/or be able to update itself to provide the latest info on the chain.

>I have not been able to try it yet because I have not been able to get truffle-contract installed. I have tried npm install truffle-contract (and same with sudo) but I didn't succeed.

What error are you seeing when trying to install `truffle-contract`?

>The const web3 is what is going to change for each node where I am going to install truffle-contract, ins't? And the ContractJSON actually needs to be define for every contract I want to interact, and for each node?

Yes, whichever node you want to use the modeler with will need to have the contract json artifacts (or access to it hosted somewhere), `web3` and the `truffle-contract` pkgs, and probably also the modeler js pkg.

>By copying those .json files into each node, I assure that they are working through the same contract instance, don't I? 

The json's in conjunction with `web3` and `truffle-contract` help make sense of what's actually happening on chain via a created contract instance. Since you'll be using two different contract instances using the same artifacts, I _think_ you'll have to make sure if one makes a call or send's a tx, the other will need to either be polling or update itself via a method call.

>If I am using truffle console to try my smart contract functions, how can I show on console the boolean that, for example, my public function "AssociateMasterDataParty" should be retrieving?

Can you share the output of the tx receipt? There might be a boolean somewhere in the logs. If not, you will probably need to `emit` an event with the bool about to be returned before properly returning the bool. Thereafter you should be able to parse the tx.logs for the returned value.

**Comment by romisalve** [comment]
@CruzMolina 
> The json's in conjunction with `web3` and `truffle-contract` help make sense of what's actually happening on chain via a created contract instance. Since you'll be using two different contract instances using the same artifacts, I _think_ you'll have to make sure if one makes a call or send's a tx, the other will need to either be polling or update itself via a method call.

Eventhough I use different instances I will be modifying the same mappings, arrays, etc? Or each structure is created for each instance? Because I need them to be the same no matter from where I access them. 
What do you refer with "method call"?



> What error are you seeing when trying to install `truffle-contract`?

![image](https://user-images.githubusercontent.com/21060405/58581281-919e9200-8225-11e9-8f38-c231528b8fda.png)



> Can you share the output of the tx receipt?
![image](https://user-images.githubusercontent.com/21060405/58581322-aa0eac80-8225-11e9-8b0d-cd051a3f1349.png)

And I am having trouble with the ReadAEO function, it is not working, it appears the same mistake of "did you ran out of gas?" 
![image](https://user-images.githubusercontent.com/21060405/58581436-ec37ee00-8225-11e9-8263-02db794fbc28.png)

The "FuncRefNum1" and "PrimerpartyNameAEO" where the values used on the createMasterDataAEO function and on associateMasterDataParty respectively.

**Comment by romisalve** [comment]
@CruzMolina truffle-contract can be on a different machine than the modeler? And with truffle contract I won't be exposing an URL to hit it and consume the smart contract methods, won't I? There is no way to have that , isn't it?

**Comment by romisalve** [comment]
@CruzMolina  now the function readAEO does not retrieve an error, but its output is giving empty strings or how this should be interpreted? 
![image](https://user-images.githubusercontent.com/21060405/58711705-60dd6a80-8395-11e9-9e9b-c8a570fc7abb.png)

It may be happening that my structures are empty?

**Comment by romisalve** [comment]
@CruzMolina is there a way to assure that both contracts use the same instance of the inherited contract? 
The way I'm doing it now, each contract is combining its own methods and the methods of the inherited one. This way both contracts will have some of the same methods and mappings and arrays, but only in name; each contract has it's own state where actual values will be store. 

Is there some way to assure both contracts are using the same instance of CrudAEOStructures? Or which would be a workaround

Thank you!!

**Comment by eggplantzzz** [comment]
truffle-contract is just a tool which provides a way to interact with contracts on chain. The main thing it consumes is the artifact files that are created by Truffle, which are currently in JSON format. So if you want to interact with the same contracts you could copy over the JSON files that are associated with the artifacts you are interested in.

One caveat, however, is that these JSON files get updated when you make changes...for example, when you edit a contract and re-compile it. I suppose you could put the JSON in a place accessible to both computers, but you would have to re-load somehow every time it changes to stay in sync.

**Comment by eggplantzzz** [comment]
Also, it looks to me like there is some permission issue on your computer where you are attempting to install truffle-contract. I think you need to have root permissions to install. Go ahead and check what permissions your particular user has.

**Comment by romisalve** [comment]
@eggplantzzz thank you!! So truffle-contract is the way to consume the smart contract methods isn't it? 
Or do you know some way to "expose" those methods on kind of an url to consume them from, suppose something as postman or something like that? 
Or that is what I am doing if I use truffle-contract? or it is impossible?

**Comment by eggplantzzz** [comment]
truffle-contract just provides a way to interface with the contracts. You would need to find some other technology to use to access them via http or somesuch. I suppose you could do something like set up an Express.js server that could deal with that kind of thing...there are countless options :)

**Comment by romisalve** [comment]
@eggplantzzz ah ok, do you have kind of an example of how would be doing that? 

And also, if I am working on a contract using an instance of another contract which has certain public variables. If one of them is a public array or a public mapping, if I want to add a new element, I should define a function on the other contract to achieve this because I can't do it directly accessing the variables,can I?

Thank you!

**Comment by eggplantzzz** [comment]
Here are the docs for express https://expressjs.com/

truffle-contract has methods for making calls to on chain stuff to get values. Check out the docs https://truffleframework.com/docs/truffle/getting-started/interacting-with-your-contracts

**Comment by romisalve** [comment]
Ok, @eggplantzzz thank you, whean I reach that stage I will investigate more on it.

Now I am facing the following problem and maybe you can give me a hand on it. I am working on a POC in which I have created three contracts which all together aims to act as a CRUD contract. The reason I have split them instead of having only one, is that when I tried to have only one contract it didn't compile due to size limitations. 
At first, I tried working with inheritance as @CruzMolina  suggested, but I had to change the approach due to the following reason. When I decided to follow the inheritance approach, I had a contract in which I define certain structres (arrays, mappings, etc) which I needed them to be modified by my other 2 contracts (which were inheriting the first one). Unfourtunately, it happens that everytime I got deployed one of this last two contracts, an instance of the first one was generated so I was not working with same mappings/arrays etc. This is why I had to change my approach to the  following one.

I changed my contracts and instead of working with inheritance, I want to work with calls to the first contract. I have a library in which I define certain structs, a contract in which I define all the mappings/arrays/etc and then 2 contracts from which I make calls to the first one in order to modify its mappings/arrays/etc. I understood that I need to define an instance of the first one, specifying the address in which it was deployed (in order to assure being working with the same mappings/arrays/etc from both contracts), on both of my other contracts. 

I am defining the instance on the contract constructor, but I believe I missunderstood how to define a constructor or I know what I am doing wrong but I am getting the following error on the deployment stage: 

_Error: Invalid number of parameters for "undefined". Got 2 expected 0!
_ 
My migration file has this statement to deploy contracts: 
_module.exports = function(deployer){

  deployer.deploy(CrudAEOStructures,{from: "0x1caFCe8fF232515Dcb07eE7901D49cFF8397eF7c"}, {overwrite: false}).then(function() {
  	return deployer.deploy(CrudAEOCreateAndRead, CrudAEOStructures.address, {from: "0x1caFCe8fF232515Dcb07eE7901D49cFF8397eF7c"});
 });
_

what does it means? It is related  to the constructor define on some of my contracts? 

this is the way I have defined both constructors: 

**constructor of the contract CRUDAEOStructures:**
_constructor() public {
       totalAEOs = 0;
   }_

**constructor of the contract CRUDAEOCreateAndRead:**
_constructor(address addressOfDeployedCrudAEOStructures) public {
       // bytecode will be loaded into memory from an existing deployment
       contractNeeded= CrudAEOStructures(addressOfDeployedCrudAEOStructures);
   }
_  

Or how should I define the instance of the contract I need to work with? 

Thank you!

**Comment by eggplantzzz** [comment]
I think it is interpreting your extra 2 arguments to the deploy function as constructor arguments. I think the second and third arguments should be just one object. So in other words, it should look like this:

```
module.exports = function(deployer){

  deployer.deploy(
    CrudAEOStructures, 
    { 
      from: "0x1caFCe8fF232515Dcb07eE7901D49cFF8397eF7c",
      overwrite: false 
    }).then(function() {
  	return deployer.deploy(
          CrudAEOCreateAndRead, 
          CrudAEOStructures.address, 
          { from: "0x1caFCe8fF232515Dcb07eE7901D49cFF8397eF7c" }
        );
 });
```

You probably already know about this documentation, but just in case you don't, check out https://www.trufflesuite.com/docs/truffle/getting-started/running-migrations#deployer-api

**Comment by romisalve** [comment]
@eggplantzzz thank you !! That was the problem, I knew the documentation but I was not being able to recognize what I was doing wrong. 
Once I correct that, I faced another issue that was that my contracts were exceeding the gas limit. I separate the library to another file and then I was able to deploy.

I was wondering if maybe do you have some example of the suggestion you made of working with a server (express.js for example) and truffle-contract to illustrate me. 

And finally, if I am using a library on my contracts, should I define an instance of it on each contract specifying the address on which it was deployed or with importing it is fine? It would be getting deployed  more than once?  
On my migrations file I have no statement on deploying the library specifically, but it gets deployed whith being imported on the contract, isn't ?

Thank you a lot!!

**Comment by romisalve** [comment]
```
function prueba() public returns (uint256 retorno){
        retorno= 8989;
return retorno;
}
```


I have defined this extremely basic function in order to test if I was being able to execute at least that of my deployed contract. Anyway, when I tried, I got
![image](https://user-images.githubusercontent.com/21060405/59299617-78550700-8c63-11e9-9036-6df72d28a238.png)

What does it refers to with those 8 expected arguments ??

Maybe there is another problem that is leading to this one.

Thank you a lot!

**Comment by eggplantzzz** [comment]
It looks strange, but with the limited amount of code above I can't really tell what is going on. Do you have your code available and can you give me very specific steps telling me what you are doing?

**Comment by romisalve** [comment]
@eggplantzzz  thank you! Yes of course, my code is here: https://github.com/romisalve/Quorum-POC.git

What I am trying to do through truffle console, is: once the contracts got deployed on the network (I ran truffle migrate --reset), I tried: 

`CrudAEOCreateAndRead.deployed().then(function(instance) { crud = instance;})`  to get a contract instance in order to be able to try contract functions. 

and then I tried `crud.createMasterDataAEO(1234, "FuncRefNum1", "primerDocName", "receipType", "receiptName", "senderType", "senderAEO", "29052019" )`  which is the first function of CrudAEOCreateAndRead.sol contract. That first function failed, so I tried to define a very basic function,  for example
```
function prueba() public returns (uint256 retorno){
        retorno= 8989;
return retorno;
}  
```
inside CrudAEOCreateAndRead.sol just to try if at least that works, and it gives me the error of the previous post.  This last function is not on my contract because in fact it doesn't do anything, I was just trying to see if something could be executed.

**Comment by romisalve** [comment]
@eggplantzzz for some reason the ABI of my contract is: 
![image](https://user-images.githubusercontent.com/21060405/59374780-ff1ce900-8d22-11e9-94e7-589d3520fd70.png)

`function prueba`  seems to be defined with 8 inputs but on the contract code is as I mentioned: 
![image](https://user-images.githubusercontent.com/21060405/59374841-27a4e300-8d23-11e9-96f0-c03039a2c986.png)

And I also do not understand why contractNeeded appears as a method if I define it as an instance (or that was my intention) 

Thank you!

**Comment by eggplantzzz** [comment]
@romisalve Sorry for the delay in getting back to you, I have been traveling for the last week. I'll take a look and get back to you soon.

**Comment by eggplantzzz** [comment]
I'm having trouble compiling your code. Your compilers section in `truffle-config.js` is in the networks section when it should be outside it. Hmm, it seems strange that the abi is like that...`createMasterDataAEO` has that signature so maybe it is somehow getting confused with that. What happens when you change the name of that method?

**Comment by romisalve** [comment]
@eggplantzzz  Sorry I am answering so late, I found that there was a problem with the truffle version. There was a bug which generates incorrect artifacts when compiling. 
I believe my code has problems due to the fact that I am using too many for loops with undefined number of iterations, or something to do with that, but I don't know how could I avoid those loops. 

I started testing every part of my function separately because I was not understanding why my transaction associated to that function was being reverted. 
I found that whenever I was pushing new elements to my array, I was not defining any int and as every push operation was retrieving an integer, that was leading my transactions to failure. Maybe that has nothing to do, but when I changed that, my transactions were no longer reverted. 
Then, once that was solved, I added the for loop and ended up facing the problem with the loops. 

I am going to update my repo with that change, but I have not came with a solution yet.

Thank you a lot!

**Comment by fainashalts** [comment]
Hi @romisalve, we're closing this for issue maintenance as its been a while. If you'd still like some guidance here, please open a new issue and we'll take a look!
